domain/consensus/processes/dagtraversalmanager/window.go

```diff
@@ -1,38 +1,80 @@
 package dagtraversalmanager
 
-import "github.com/kaspanet/kaspad/domain/consensus/model/externalapi"
+import (
+	"sort"
+
+	"github.com/kaspanet/kaspad/domain/consensus/model/externalapi"
+)
 
 // blueBlockWindow returns a blockWindow of the given size that contains the
 // blues in the past of startindNode, sorted by GHOSTDAG order.
 // If the number of blues in the past of startingNode is less then windowSize,
 // the window will be padded by genesis blocks to achieve a size of windowSize.
-func (dtm *dagTraversalManager) BlueWindow(startingBlock *externalapi.DomainHash, windowSize uint64) ([]*externalapi.DomainHash, error) {
-	window := make([]*externalapi.DomainHash, 0, windowSize)
-
+func (dtm *dagTraversalManager) BlueWindow(startingBlock *externalapi.DomainHash, windowSize int) ([]*externalapi.DomainHash, error) {
 	currentHash := startingBlock
 	currentGHOSTDAGData, err := dtm.ghostdagDataStore.Get(dtm.databaseContext, currentHash)
 	if err != nil {
 		return nil, err
 	}
 
-	for uint64(len(window)) < windowSize && currentGHOSTDAGData.SelectedParent != nil {
-		for _, blue := range currentGHOSTDAGData.MergeSetBlues {
-			window = append(window, blue)
-			if uint64(len(window)) == windowSize {
+	windowHeap := dtm.newSizedUpHeap(windowSize)
+
+	for windowHeap.len() <= windowSize && currentGHOSTDAGData.SelectedParent != nil {
+		added, err := windowHeap.tryPush(currentGHOSTDAGData.SelectedParent)
+		if err != nil {
+			return nil, err
+		}
+
+		// If the window is full and the selected parent is less than the minimum then we break
+		// because this means that there cannot be any more blocks in the past with higher blueWork
+		if !added {
+			break
+		}
+
+		// Now we go over the merge set.
+		// Remove the SP from the blue merge set because we already added it.
+		currentGHOSTDAGData.MergeSetBlues = currentGHOSTDAGData.MergeSetBlues[1:]
+		// Go over the merge set in reverse because it's ordered in reverse by blueWork.
+		for i := len(currentGHOSTDAGData.MergeSetBlues) - 1; i >= 0; i-- {
+			added, err := windowHeap.tryPush(currentGHOSTDAGData.MergeSetBlues[i])
+			if err != nil {
+				return nil, err
+			}
+			// If it's smaller than minimum then we won't be able to add the rest because they're even smaller.
+			if !added {
+				break
+			}
+		}
+		for i := len(currentGHOSTDAGData.MergeSetReds) - 1; i >= 0; i-- {
```

Comment:
There can be a situation where a red has higher blue work than a blue in the merge set, but it won't be added because the window is full

Reply:
That can't happen because they are separate loops and the `break` only breaks the first loop